File: Algorithms/235. Lowest Common Ancestor of a Binary Search Tree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn lowest_common_ancestor(root: Option<Rc<RefCell<TreeNode>>>, p: Option<Rc<RefCell<TreeNode>>>, q: Option<Rc<RefCell<TreeNode>>>) -> Option<Rc<RefCell<TreeNode>>> {
        let pVal = p.clone().unwrap().borrow().val;
        let qVal = q.clone().unwrap().borrow().val;
        let pNode = root.clone().unwrap().borrow().val;
        if pVal < pNode && qVal < pNode {
            return Self::lowest_common_ancestor(root.unwrap().borrow().left.clone(), p, q);
        } else if pVal > pNode && qVal > pNode {
            return Self::lowest_common_ancestor(root.unwrap().borrow().right.clone(), p, q);
        } else {
            return root;
        }
    }
}

File: Algorithms/235. Lowest Common Ancestor of a Binary Search Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function lowestCommonAncestor(root: TreeNode | null, p: TreeNode | null, q: TreeNode | null): TreeNode | null {

    const dfs = (root: TreeNode | null, p: TreeNode | null, q: TreeNode | null) => {
        if ([root, p, q].includes(null)) return root;

        if (p.val > root.val && q.val > root.val)
            return dfs(root.right, p, q)

        if (p.val < root.val && q.val < root.val)
            return dfs(root.left, p, q);

        return root;
    }
    return dfs(root, p, q);
};
